File: src/java/foundgine-core/src/main/java/com/foundgine/core/execution/mutation/MutationDependencyGraph.java
package com.foundgine.core.execution.mutation;

import com.foundgine.core.semantic.planning.mutation.MutationDependency;

import java.util.LinkedHashSet;
import java.util.List;
import java.util.Objects;
import java.util.Set;

/**
 * Canonical provider-neutral execution dependency graph. Dependency edges are
 * already resolved to operation ordinals and physical target columns.
 * Provider-specific correlation carriers are introduced only during provider
 * lowering.
 */
public final class MutationDependencyGraph {

	/**
	 * A directed edge between two operation indexes in the mutation dependency
	 * graph.
	 */
	public record Edge(int sourceOperationIndex, int targetOperationIndex) {
	}

	private final List<MutationDependency> dependencies;

	public MutationDependencyGraph(List<MutationDependency> dependencies) {
		Objects.requireNonNull(dependencies);
		this.dependencies = List.copyOf(dependencies);
	}

	public List<MutationDependency> dependencies() {
		return dependencies;
	}

	public Set<Edge> edges() {
		Set<Edge> edges = new LinkedHashSet<>();
		for (MutationDependency d : dependencies)
			edges.add(new Edge(d.sourceOperationIndex(), d.targetOperationIndex()));
		return edges;
	}
}

File: src/csharp/Foundgine.Core/Execution/Mutation/MutationDependencyGraph.cs
using Foundgine.Core.Semantic.Planning.Mutation;

namespace Foundgine.Core.Execution.Mutation;

/// <summary>
/// Canonical provider-neutral execution dependency graph.
/// Dependency edges are already resolved to operation ordinals and physical
/// target columns. Provider-specific correlation carriers are introduced only
/// during provider lowering.
/// </summary>
public sealed class MutationDependencyGraph
{
    private readonly IReadOnlyList<MutationDependency> _dependencies;

    public IReadOnlyList<MutationDependency> Dependencies => _dependencies;

    public MutationDependencyGraph(IEnumerable<MutationDependency> dependencies)
    {
        ArgumentNullException.ThrowIfNull(dependencies);
        _dependencies = dependencies.ToArray();
    }

    public IReadOnlySet<(int SourceOperationIndex, int TargetOperationIndex)> Edges =>
        _dependencies
            .Select(d => (d.SourceOperationIndex, d.TargetOperationIndex))
            .ToHashSet();
}